tdrs-backend/tdpservice/search_indexes/parsers/tanf_parser.py

```diff
@@ -0,0 +1,100 @@
+"""Transforms a TANF datafile into an search_index model."""
+
+import logging
+from ..models import T1  # , T2, T3, T4, T5, T6, T7, ParserLog
+# from django.core.exceptions import ValidationError
+from .util import get_record_type
+from .schema_defs.tanf import t1_schema
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+def active_t1_parser(line, line_number):
+    """Parse line in datafile as active case data, T1 only."""
+    family_case_schema = t1_schema()
+    # create search_index model
+    t1 = T1()
+    content_is_valid = True
+    for field in family_case_schema.get_all_fields():
+        if field.name == 'blank':
+            break
+        content = line[field.start-1:field.end]  # descriptor pdfs were off by one, could also adjust start values
+        if len(content) != field.length:
+            logger.warn('[LineNo:%d, col%d] Expected field "%s" with length %d, got: "%s"',
+                        line_number, field.start-1, field.name, field.length, content)
+            content_is_valid = False
+            continue
+        # check if content is type string or integer
+        if field.type == 'Numeric':
```

Comment:
this is where i think we can start to abstract these checks into different functions. i'm envisioning something like this

```python
def numeric_field_validator():
    # impl

def alphanumeric_field_validator():
    # impl

validators = {
    'num_field': [
        numeric_field_validator,
        greater_than_validator(7)
    ],
    'char_field': [
        alphanumeric_field_validator,
        options_validator(['0', '1', '2', '3', '4')
    ]
}
```

def active_t1_parser():
    # does all the stuff 
    for field in schema:
        content = line[field.start-1:field.end]
        validator_funcs = validators[field]
        for validator in validator_funcs:
            validator(value) # this should work if each validator works like a pure function. we could potentially include validators that relate to other fields in the model as well

Reply:
Great idea for #1354